notebooks/notebook_template_review.py

```diff
@@ -154,268 +157,354 @@ class ErrorCode(Enum):
+    @property
+    def path(self):
+        '''
+        Getter: return the filename path for the notebook
+        '''
+        return self._path
+    
 
-        InstallationRule().validate(path, cells)
+    @property
+    def num_errors(self):
+        '''
+        Getter: return the number of errors
+        '''
+        return self._num_errors
 
-        RestartRule().validate(path, cells)
-        
-        VersionsRule().validate(path, cells)
         
-        BeforeBeginRule().validate(path, cells)
+    def report_error(self,
+                     code: ErrorCode,
+                     errmsg: str):
+        """
+        Report an error.
+            If args.errors_codes set, then only report these errors. Otherwise, all errors.
+
+        code: The error code number.
+        errmsg: The error message
+        """
+
+        if args.errors:
+            code = code.value[0]
+            if args.errors_codes:
+                if str(code) not in args.errors_codes:
+                    return
+
+            if args.errors_csv:
+                print(self._path, ',', code)
+            else:
+                print(f"{self._path}: ERROR ({code}): {errmsg}", file=sys.stderr)
+                self._num_errors += 1
+                
+            return False
+        return True
+
+
+    def report_fix(self,
+                   code: ErrorCode,
+                   fixmsg: str):
+        """
+        Report an automatic fix
         
-        EnableAPIsRule().validate(path, cells)
+        code: The error code number.
+        fixmsg: The autofix message
+        """
+        print(f"{self._path}: FIXED ({code}): {fixmsg}", file=sys.stderr)
         
-        SetupProjectRule().validate(path, cells)
+                
+    def writeback(self):
+        """
+        Write the updated (autofixed) notebook to a temporary file
+        """
+        with open('tmp.ipynb', 'w') as f:
+            json.dump(self._content, f)
+
 
 
 class NotebookRule(ABC):
     """
     Abstract class for defining notebook conformance rules
     """
-    cell_index = 0
-    title = ''
-    git_link = None
-    colab_link = None
-    workbench_link = None
-    desc = ''
-    uses = ''
-    steps = ''
-    costs = []
-    
-    @staticmethod
-    def init():
-        NotebookRule.cell_index = 0
-        NotebookRule.title = ''
-        NotebookRule.git_link = None
-        NotebookRule.colab_link = None
-        NotebookRule.workbench_link = None
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
-        pass
+    def validate(self, notebook: Notebook) -> bool:
+        '''
+        Applies cell specific rules to validate whether the cell 
+        does or does not conform to the rules.
+        
+        Returns whether the cell passed the validation rules
+        '''
+        return False
```

Comment:
This is an abstract method, so should be `pass`

Reply:
done